### PR TITLE
docs on how to optimize dogstatsd for high throughput

### PR DESCRIPTION
**Pull request by arbll**
### What does this PR do?
Adds some tips on how to optimize dogstatsd to be able to handle large volume of metrics.

### Motivation
Help users configure their agent/dogstatsd client/kernel when they reach the limits of their default configuration.

### Preview link (Pipeline is broken right now)
https://docs-staging.datadoghq.com/arbll/dogstatsd-high-throughput/developers/dogstatsd/high_throughput/

### CODE REVIEW

**Comment by l0k0ms** [comment]
@arbll i merged master to fix a conflict and the pre-hook job changed some link reference, don't mind the 3 extra files in your PR.

File: content/en/developers/dogstatsd/high_throughput.md [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+// packets containing several metrics
+$client->increment('your.data.point');
+$client->increment('your.data.point', .5);
+```
+{{% /tab %}}
+{{< /tabs >}}
+
+### Sample your metrics
+
+It is possible to reduce the traffic from your DogStatsD client and the Agent by setting a sample
```
Comment (KSerrania) [inline content/en/developers/dogstatsd/high_throughput.md]: Either:
`from your DogStatsD client to the Agent`
or
`between your DogStatsD client and the Agent`

File: content/en/developers/dogstatsd/high_throughput.md [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+rate value for your client. A sample rate of 0.5 cuts the number of UDP packets sent in half.
```
Comment (KSerrania) [inline content/en/developers/dogstatsd/high_throughput.md]:
```suggestion
rate value for your client. A sample rate of 0.5 cuts the number of UDP packets sent by half.
```

File: content/en/developers/dogstatsd/high_throughput.md [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+metrics will start getting dropped.
+
+It is possible to adjust those values to give the agent more time to process
+incoming metrics.
+
+### Over UDP (User Datagram Protocol)
+
+#### Linux
+
+On most Linux distribution the maximum size of the kernel buffer is set to 212992
```
Comment (KSerrania) [inline content/en/developers/dogstatsd/high_throughput.md]:
```suggestion
On most Linux distributions the maximum size of the kernel buffer is set to 212992
```

File: content/en/developers/dogstatsd/high_throughput.md [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+```bash
+$ sysctl -w net.core.rmem_max=26214400
+```
+
+Add the following to `/etc/sysctl.conf` to make this change permanent:
+```
+net.core.rmem_max = 26214400
+```
+
+Set the agent `dogstatsd_so_rcvbuf` confing option to the same number in `datadog.yaml`:
```
Comment (KSerrania) [inline content/en/developers/dogstatsd/high_throughput.md]:
```suggestion
Set the Agent `dogstatsd_so_rcvbuf` config option to the same number in `datadog.yaml`:
```

File: content/en/developers/dogstatsd/high_throughput.md [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+It has very little overhead when compared to UDP and will lower the
+general footprint of dogstatsd on your system.
+
+## Operating System kernel buffers
+
+Most OSs will add incoming UDP and UDS datagrams containing your metrics to a
+buffer with a maximum size. Once this size is reached, datagrams containing your
+metrics will start getting dropped.
+
+It is possible to adjust those values to give the agent more time to process
```
Comment (KSerrania) [inline content/en/developers/dogstatsd/high_throughput.md]:
```suggestion
It is possible to adjust those values to give the Agent more time to process
```

File: content/en/developers/dogstatsd/high_throughput.md [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+
+Most of the time those symptoms can be alleviated by tweaking some configuration
+options described below.
+
+## General tips
+
+### Enable buffering on your client
+
+Some statsd/dogstatsd clients, by default, will send one metric per datagram.
+This adds considerable overhead on the client, the operating system and the agent.
```
Comment (KSerrania) [inline content/en/developers/dogstatsd/high_throughput.md]:
```suggestion
This adds considerable overhead on the client, the operating system and the Agent.
```

File: content/en/developers/dogstatsd/high_throughput.md [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+If your client supports buffering multiple metrics in one datagram enabling this
```
Comment (KSerrania) [inline content/en/developers/dogstatsd/high_throughput.md]:
```suggestion
If your client supports buffering multiple metrics in one datagram, enabling this
```

File: content/en/developers/dogstatsd/high_throughput.md [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+- link: "https://github.com/DataDog/datadog-agent/tree/master/pkg/dogstatsd"
+  tag: "GitHub"
+  text: "DogStatsD source code"
+---
+
+DogStatsD works by sending metrics generated from your application to the
+[agent](https://docs.datadoghq.com/agent/) over a transport protocol. This
+transport protocol can be either UDP (User Datagram Protocol) or UDS (Unix Domain Socket).
+
+When dogstatsd is used to send a large volume of metrics to a single agent if
```
Comment (KSerrania) [inline content/en/developers/dogstatsd/high_throughput.md]:
```suggestion
When DogStatsD is used to send a large volume of metrics to a single Agent if
```

File: content/en/developers/dogstatsd/high_throughput.md [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+- (UDS) DogStatsD client library returning lots of errors
+
+Most of the time those symptoms can be alleviated by tweaking some configuration
+options described below.
+
+## General tips
+
+### Enable buffering on your client
+
+Some statsd/dogstatsd clients, by default, will send one metric per datagram.
```
Comment (KSerrania) [inline content/en/developers/dogstatsd/high_throughput.md]:
```suggestion
Some statsd/DogStatsD clients, by default, will send one metric per datagram.
```

File: content/en/developers/dogstatsd/high_throughput.md [hunk 10]
```diff
@@ -0,0 +1,199 @@
...
+rate value for your client. A sample rate of 0.5 cuts the number of UDP packets sent in half.
+This solution is a trade-off: you decrease traffic but slightly lose in precision/granularity.
+
+A more in-depth explanation with code examples is available
+[here](https://docs.datadoghq.com/developers/faq/dog-statsd-sample-rate-parameter-explained/).
+
+### Use DogStatsD over UDS (Unix Domain Socket)
+
+UDS is an inter-process communication protocol that can
+be [used to transport dogstatsd payloads](https://docs.datadoghq.com/developers/dogstatsd/unix_socket/).
```
Comment (KSerrania) [inline content/en/developers/dogstatsd/high_throughput.md]:
```suggestion
be [used to transport DogStatsD payloads](https://docs.datadoghq.com/developers/dogstatsd/unix_socket/).
```

File: content/en/developers/dogstatsd/high_throughput.md [hunk 11]
```diff
@@ -0,0 +1,199 @@
...
+If your client supports buffering multiple metrics in one datagram enabling this
+option could bring noticeable improvements.
+
+Here are a few examples for supported clients:
+
+Please refer to your client documentation for additional details.
+
+{{< tabs >}}
+{{% tab "Go" %}}
+Using Datadog's official Golang library [datadog-go](https://github.com/DataDog/datadog-go):
```
Comment (apigirl) [inline content/en/developers/dogstatsd/high_throughput.md]:
```suggestion
Use Datadog's official Golang library [datadog-go](https://github.com/DataDog/datadog-go):
```

File: content/en/developers/dogstatsd/high_throughput.md [hunk 12]
```diff
@@ -0,0 +1,199 @@
...
+    statsd.WithMaxMessagesPerPayload(256), // sets the maximum number of messages in a single datagram
+)
+// All metrics sent from this instance of the client will be buffered and sent in
+// packets containing several metrics
+client.Gauge("kafka.health", 1, []string{"env:production", "partition:1", "partition:2"}, 1)
+```
+{{% /tab %}}
+
+{{% tab "Python" %}}
+Using Datadog's official Python library [datadogpy](https://github.com/DataDog/datadogpy):
```
Comment (apigirl) [inline content/en/developers/dogstatsd/high_throughput.md]:
```suggestion
Use Datadog's official Python library [datadogpy](https://github.com/DataDog/datadogpy):
```

File: content/en/developers/dogstatsd/high_throughput.md [hunk 13]
```diff
@@ -0,0 +1,199 @@
...
+
+# All metrics will be buffered and sent in one packet when the block completes
+with DogStatsd(host="127.0.0.1", port=8125, max_buffer_size=25) as batch:
+    batch.gauge('users.online', 123)
+    batch.gauge('active.connections', 1001)
+```
+{{% /tab %}}
+
+{{% tab "Ruby" %}}
+Using Datadog's official Ruby library [dogstatsd-ruby](https://github.com/DataDog/dogstatsd-ruby):
```
Comment (apigirl) [inline content/en/developers/dogstatsd/high_throughput.md]:
```suggestion
Use Datadog's official Ruby library [dogstatsd-ruby](https://github.com/DataDog/dogstatsd-ruby):
```

File: content/en/developers/dogstatsd/high_throughput.md [hunk 14]
```diff
@@ -0,0 +1,199 @@
...
+# All metrics will be buffered and sent in one packet when the block completes
+statsd.batch do |s|
+  s.increment('page.views')
+  s.gauge('users.online', 123)
+end
+```
+{{% /tab %}}
+
+{{% tab "Java" %}}
+Using Datadog's official Java library [java-dogstatsd-client](https://github.com/DataDog/java-dogstatsd-client):
```
Comment (apigirl) [inline content/en/developers/dogstatsd/high_throughput.md]:
```suggestion
Use Datadog's official Java library [java-dogstatsd-client](https://github.com/DataDog/java-dogstatsd-client):
```

File: content/en/developers/dogstatsd/high_throughput.md [hunk 15]
```diff
@@ -0,0 +1,199 @@
...
+// packets containing several metrics
+public void foo() {
+    statsd.incrementCounter("foo");
+    statsd.recordGaugeValue("bar", 100);
+}
+```
+{{% /tab %}}
+
+{{% tab "C#" %}}
+Using Datadog's official C# library [dogstatsd-csharp-client](https://github.com/DataDog/dogstatsd-csharp-client):
```
Comment (apigirl) [inline content/en/developers/dogstatsd/high_throughput.md]:
```suggestion
Use Datadog's official C# library [dogstatsd-csharp-client](https://github.com/DataDog/dogstatsd-csharp-client):
```

File: content/en/developers/dogstatsd/high_throughput.md [hunk 16]
```diff
@@ -0,0 +1,199 @@
...
+    s.Add<Statsd.Counting,int>("stat-name", 1);
+
+    // All metrics buffered before this call will be sent in one packet
+    s.Send();
+}
+```
+{{% /tab %}}
+
+{{% tab "PHP" %}}
+Using Datadog's official PHP library [php-datadogstatsd](https://github.com/DataDog/php-datadogstatsd):
```
Comment (apigirl) [inline content/en/developers/dogstatsd/high_throughput.md]:
```suggestion
Use Datadog's official PHP library [php-datadogstatsd](https://github.com/DataDog/php-datadogstatsd):
```

File: content/en/developers/dogstatsd/high_throughput.md (hunk 1 above)
Comment (apigirl) [inline content/en/developers/dogstatsd/high_throughput.md]:
```suggestion
It is possible to reduce the traffic from your DogStatsD client to the Agent by setting a sample
```

File: content/en/developers/dogstatsd/high_throughput.md [hunk 17]
```diff
@@ -0,0 +1,199 @@
...
+
+On most Linux distribution the maximum size of the kernel buffer is set to 212992
+by default (208 KiB). This can be confirmed using the following commands:
+
+```bash
+$ sysctl net.core.rmem_max
+net.core.rmem_max = 212992
+```
+
+To set the maximum size of the dogstatsd socket buffer to 25MiB:
```
Comment (apigirl) [inline content/en/developers/dogstatsd/high_throughput.md]:
```suggestion
To set the maximum size of the DogStatsD socket buffer to 25MiB:
```

**Comment by arbll** [comment]
Thanks for the review @apigirl and @KSerrania!

@apigirl AFAIK markdown allows for such line breaks to be used to make the unrendered text more readable:

> A paragraph is simply one or more consecutive lines of text, separated by one or more blank lines. (A blank line is any line that looks like a blank line — a line containing nothing but spaces or tabs is considered blank.) Normal paragraphs should not be indented with spaces or tabs.

> The implication of the “one or more consecutive lines of text” rule is that Markdown supports “hard-wrapped” text paragraphs. This differs significantly from most other text-to-HTML formatters (including Movable Type’s “Convert Line Breaks” option) which translate every line break character in a paragraph into a <br /> tag.

> When you do want to insert a <br /> break tag using Markdown, you end a line with two or more spaces, then type return.

> Yes, this takes a tad more effort to create a <br />, but a simplistic “every line break is a <br />” rule wouldn’t work for Markdown. Markdown’s email-style blockquoting and multi-paragraph list items work best — and look better — when you format them with hard breaks. 
https://daringfireball.net/projects/markdown/syntax#p

But I will remove those line breaks since it seems to be consistent with the rest of the repo

File: content/en/developers/dogstatsd/high_throughput.md [hunk 18]
```diff
@@ -0,0 +1,184 @@
...
+  text: "Introduction to DogStatsD"
+- link: "developers/libraries"
+  tag: "Documentation"
+  text: "Official and Community-contributed API and DogStatsD client libraries"
+- link: "https://github.com/DataDog/datadog-agent/tree/master/pkg/dogstatsd"
+  tag: "GitHub"
+  text: "DogStatsD source code"
+---
+
+DogStatsD works by sending metrics generated from your application to the [agent](https://docs.datadoghq.com/agent/) over a transport protocol. This transport protocol can be either UDP (User Datagram Protocol) or UDS (Unix Domain Socket).
```
Comment (KSerrania) [inline content/en/developers/dogstatsd/high_throughput.md]:
```suggestion
DogStatsD works by sending metrics generated from your application to the [Agent](https://docs.datadoghq.com/agent/) over a transport protocol. This transport protocol can be either UDP (User Datagram Protocol) or UDS (Unix Domain Socket).
```

File: content/en/developers/dogstatsd/high_throughput.md [hunk 19]
```diff
@@ -0,0 +1,184 @@
...
+
+## General tips
+
+### Enable buffering on your client
+
+Some StatsD and DogStatsD clients, by default, will send one metric per datagram. This adds considerable overhead on the client, the operating system, and the Agent. If your client supports buffering multiple metrics in one datagram, enabling this option could bring noticeable improvements.
+
+Here are a few examples for supported clients:
+
+Please refer to your client documentation for additional details.
```
Comment (KSerrania) [inline content/en/developers/dogstatsd/high_throughput.md]: I'd put this after the tabs, seems more natural (I expect the examples to be right after the sentence which introduces them)

File: content/en/developers/dogstatsd/high_throughput.md [hunk 20]
```diff
@@ -0,0 +1,184 @@
...
+// packets containing several metrics
+$client->increment('your.data.point');
+$client->increment('your.data.point', .5);
+```
+{{% /tab %}}
+{{< /tabs >}}
+
+### Sample your metrics
+
+It is possible to reduce the traffic from your DogStatsD client to the Agent by setting a sample rate value for your client. For example, a sample rate of `0.5` halves number of UDP packets sent. This solution is a trade-off: you decrease traffic but slightly lose in precision and granularity.
```
Comment (KSerrania) [inline content/en/developers/dogstatsd/high_throughput.md]:
```suggestion
It is possible to reduce the traffic from your DogStatsD client to the Agent by setting a sample rate value for your client. For example, a sample rate of `0.5` halves the number of UDP packets sent. This solution is a trade-off: you decrease traffic but slightly lose in precision and granularity.
```

**Comment by arbll** [comment]
@apigirl Could you give this another look ?

**Pull request merged**